src/Database/LSMTree/Internal/BlobRef.hs

```diff
@@ -89,82 +116,98 @@ newtype WeakBlobRefInvalid = WeakBlobRefInvalid Int
 
 {-# SPECIALISE withWeakBlobRefs ::
      V.Vector (WeakBlobRef IO h)
-  -> (V.Vector (RawBlobRef IO h) -> IO a)
+  -> (V.Vector (StrongBlobRef IO h) -> IO a)
   -> IO a #-}
 -- | The same as 'withWeakBlobRef' but for many references in one go.
 --
 withWeakBlobRefs ::
      (MonadMask m, PrimMonad m)
   => V.Vector (WeakBlobRef m h)
-  -> (V.Vector (RawBlobRef m h) -> m a)
+  -> (V.Vector (StrongBlobRef m h) -> m a)
   -> m a
 withWeakBlobRefs wrefs = bracket (deRefWeakBlobRefs wrefs) removeReferences
 
 {-# SPECIALISE deRefWeakBlobRef ::
      WeakBlobRef IO h
-  -> IO (RawBlobRef IO h) #-}
+  -> IO (StrongBlobRef IO h) #-}
 deRefWeakBlobRef ::
      (MonadThrow m, PrimMonad m)
   => WeakBlobRef m h
-  -> m (RawBlobRef m h)
-deRefWeakBlobRef (WeakBlobRef ref) = do
-    ok <- RC.upgradeWeakReference (blobFileRefCounter (rawBlobRefFile ref))
+  -> m (StrongBlobRef m h)
+deRefWeakBlobRef WeakBlobRef{weakBlobRefFile, weakBlobRefSpan} = do
+    ok <- RC.upgradeWeakReference (blobFileRefCounter weakBlobRefFile)
     when (not ok) $ throwIO (WeakBlobRefInvalid 0)
-    pure ref
+    return StrongBlobRef{
+      strongBlobRefFile = weakBlobRefFile,
+      strongBlobRefSpan = weakBlobRefSpan
+    }
 
 {-# SPECIALISE deRefWeakBlobRefs ::
      V.Vector (WeakBlobRef IO h)
-  -> IO (V.Vector (RawBlobRef IO h)) #-}
+  -> IO (V.Vector (StrongBlobRef IO h)) #-}
 deRefWeakBlobRefs ::
     forall m h.
      (MonadMask m, PrimMonad m)
   => V.Vector (WeakBlobRef m h)
-  -> m (V.Vector (RawBlobRef m h))
+  -> m (V.Vector (StrongBlobRef m h))
 deRefWeakBlobRefs wrefs = do
-    let refs :: V.Vector (RawBlobRef m h)
-        refs = coerce wrefs -- safely coerce away the newtype wrappers
-    V.iforM_ wrefs $ \i (WeakBlobRef ref) -> do
-      ok <- RC.upgradeWeakReference (blobFileRefCounter (rawBlobRefFile ref))
-      when (not ok) $ do
-        -- drop refs on the previous ones taken successfully so far
-        V.mapM_ removeReference (V.take i refs)
-        throwIO (WeakBlobRefInvalid i)
-    pure refs
-
-{-# SPECIALISE removeReference :: RawBlobRef IO h -> IO () #-}
-removeReference :: (MonadMask m, PrimMonad m) => RawBlobRef m h -> m ()
-removeReference = BlobFile.removeReference . rawBlobRefFile
-
-{-# SPECIALISE removeReferences :: V.Vector (RawBlobRef IO h) -> IO () #-}
-removeReferences :: (MonadMask m, PrimMonad m) => V.Vector (RawBlobRef m h) -> m ()
+    refs <- VM.new (V.length wrefs)
+    V.iforM_ wrefs $ \i WeakBlobRef {weakBlobRefFile, weakBlobRefSpan} -> do
+      ok <- RC.upgradeWeakReference (blobFileRefCounter weakBlobRefFile)
+      if ok
+        then VM.write refs i StrongBlobRef {
+               strongBlobRefFile = weakBlobRefFile,
+               strongBlobRefSpan = weakBlobRefSpan
+             }
+        else do
+          -- drop refs on the previous ones taken successfully so far
+          VM.mapM_ removeReference (VM.take i refs)
+          throwIO (WeakBlobRefInvalid i)
+    V.unsafeFreeze refs
```

Comment:
Is going through a mutable vector necessary, or would `V.iforM_ wrefs ...` be sufficient?

Reply:
At this stage that would probably work, but once we distinguish a `WeakRef BlobFile` from a `Ref BlobFile` then we have to release the actual `Ref BlobFile` and not just fiddle the refcount of the `WeakRef`. Otherwise we'll be breaking through the `Ref` abstraction.